### PR TITLE
rand upgrade to v0.7

### PR DESCRIPTION
**Pull request by jeanphilippeD**
Supersede https://github.com/maidsafe/routing/pull/1971

### CODE REVIEW

File: src/rng.rs [hunk 1]
```diff
@@ -50,6 +50,24 @@ impl<R: rand::Rng> rand_parsec::Rng for RngParsecCompat<R> {
...
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        self.0.fill_bytes(dest)
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand::Error> {
+        self.0.try_fill_bytes(dest)
+    }
+}
+
```
Comment (madadam) [inline src/rng.rs]: Why do we need this? Isn't this type only passed to parsec and this require only `rand_parsec::Rng` impl?

File: tests/mock_network/utils.rs [hunk 2]
```diff
@@ -830,9 +833,17 @@ pub fn verify_invariant_for_all_nodes(network: &Network, nodes: &mut [TestNode])
     );
 }
 
+// Generate a vector of random bytes of the given length.
```
Comment (madadam) [inline tests/mock_network/utils.rs]: The function is generic, so not just bytes.

File: tests/mock_network/churn.rs [hunk 3]
```diff
@@ -72,14 +72,13 @@ fn remove_unresponsive_node() {
     // Sending some user events to create a sequence of observations.
     let mut responded = 0;
     for i in 0..UNRESPONSIVE_WINDOW {
-        let event: Vec<_> = rng.gen_iter().take(100).collect();
+        let event: Vec<_> = gen_vec(&mut rng, 100);
```
Comment (madadam) [inline tests/mock_network/churn.rs]: I think the `Vec<_>` qualification is redundant now (this applies to other places in this file).

File: src/rng.rs [hunk 4]
```diff
@@ -14,19 +14,17 @@ pub use self::seed_printer::SeedPrinter;
 #[cfg(any(test, feature = "mock_base"))]
 pub use self::test::Seed;
 
-use rand::Rng;
-
-// `CryptoRng` trait shim.
+// `Cryptorand::Rng` trait shim.
```
[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/rng.rs]: Yes, but also code + comment removed in later commit.

File: src/rng.rs [hunk 5]
```diff
@@ -223,12 +220,8 @@ mod test {
         }
     }
 
-    impl Rand for Seed {
-        fn rand<R: Rng>(rng: &mut R) -> Self {
-            // Note: the `wrapping_add` trick is a workaround for what seems to be a weakness in
-            // `XorShiftRng`. Without it we would sometimes end up with multiple rngs producing
-            // identical values.
-            // The idea is taken from: https://github.com/maidsafe/maidsafe_utilities/blob/24dfcbc6ee07a14bf64f3bc573f68cea01e06862/src/seeded_rng.rs#L92
```
[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/rng.rs]: Agree, it was re-added in later commit.

File: src/xor_space/xor_name.rs [hunk 6]
```diff
@@ -262,14 +263,16 @@ impl AsRef<XorName> for XorName {
...
     use maidsafe_utilities::serialisation::{deserialise, serialise};
-    use rand;
+    use rand::Rng;
     use std::cmp::Ordering;
 
     #[test]
-    fn serialisation_xor_name() {
-        let obj_before: XorName = rand::random();
-        let data = unwrap!(serialise(&obj_before));
+    fn serialisation_xor_name() -> Result<(), RoutingError> {
```
[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/xor_space/xor_name.rs]: Agree, it was removed in later commit.

File: src/xor_space/xor_name.rs [hunk 7]
```diff
@@ -262,14 +263,16 @@ impl AsRef<XorName> for XorName {
...
+        let mut rng = rand::thread_rng();
+        let obj_before: XorName = rng.gen::<XorName>();
```
[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/xor_space/xor_name.rs]: Yes, also wrong in some placed leading to compilation error. removed in later commit.

File: src/rng.rs (hunk 1 above)
Comment (jeanphilippeD) [reply src/rng.rs]: It is also used in mock_parsec :-)

Comment (madadam) [reply src/rng.rs]: Ah, of course :)

**Comment by jeanphilippeD** [comment]
Addressed review comments and fixed commit comment.

**Comment by jeanphilippeD** [comment]
Push force again to address issue with rust format action:
```
Run actions-rs/cargo@v1
Added matchers: 'rust'. Problem matchers scan action output for known warning or error strings and report these inline.
/usr/share/rust/.cargo/bin/cargo fmt --all -- --check
Error during execution of `cargo metadata`: error: failed to download `utf8-ranges v1.0.4`
```

**Pull request merged**